Dye, Colored blocks and many fixes.

Claycorp opened the pull request:
Don't question code binges. It's a bit of a mix but it all should work well.

Main focus was new Dye system and implementing it correctly, All supporting recipes should be there.
Secondary was clearing as many simple issues off the issue list and a few direct from players.

Rip it up fam. (I think I did this all right?)

Fixes #891

Claycorp committed:
TFC Dye fluids and the recipes to go with them!

Claycorp committed:
Oops, forgot the mix recipes.

Claycorp committed:
Added crushing recipes for dye.
Renamed specialty named dye items to generic names.
Added new dye items to replace speciality dye items. (waiting on textures)
Added new Aggregate block. Just a precursor for concrete. (waiting on textures)
Added new Oredict entries to all classic dye items.
Added wooden bucket support for dyes as it would be hard to mix them otherwise.
Fixed incorrect oredict for lapis being used in the quern recipe. (Closes #897)
Fixed incorrectly defined faces for large berry bushes. Fence likes won't connect anymore. (Closes #887)

Claycorp committed:
Changed Fence gates to give 2 instead of one. (Closes TerraFirmaCraft#746)
Added unfired brick and most supporting recipes. (Closes TerraFirmaCraft#859)
(Below was fixed in the last commit but I fucked up the link stuff)
Fixed incorrect oredict for lapis being used in the quern recipe. (Closes TerraFirmaCraft#897)
Fixed incorrectly defined faces for large berry bushes. Fence likes won't connect anymore. (Closes TerraFirmaCraft#887)

Claycorp committed:
Fixed ingots stacking on thin air when using another existing stack with another type. (Closes #889)
Fixed Horse meat being one pixel off. (Thanks EERussianguy for textures Closes#776)
Almost finished adding Brick related things. (Thanks Stone for texture)

Claycorp committed:
Fixed ingots stacking on thin air when using another existing stack with another type. (Closes #889)
Fixed Horse meat being one pixel off. (Thanks EERussianguy for textures Closes#776)
Almost finished adding Brick related things. (Thanks Stone for texture)
Added Auto plan select to anvil for items with a single plan.
Added Fire Clay block for storage and decoration. (Thanks Stone for texture)
Added Aggregate, that is a cement precursor that acts like gravel. (Thanks Stone for the texture)
Added Config option to disable TFC lava cobble gen override so other mods can handle it. (Closes #849)
Added Config option to modify leaf density. (Closes #896)
Added Config option to disable TFC torch overrides so other mods can handle it. (Closes #607)
Added Config option to disable TFC tree felling so other mods can be used. (Closes #703)
Added Flower pot to follow usual clay shaping and cooking (Closes 891#)
Trapdoors now give 4 trapdoors instead to balance cost against regular doors. (Closes #848)
The end of the dye stuff! Yay!

Claycorp committed:
Needed the recipes for the new blocks...Doh.
Is this the end? PR time to get roasted.

Claycorp committed:
Fixed grass landing particales not being made of grass. (Closes #710)

alcatrazEscapee commented on src/main/java/net/dries007/tfc/CommonEventHandler.java:
```diff
@@ -685,7 +685,8 @@ public static void onFluidPlaceBlock(BlockEvent.FluidPlaceBlockEvent event)
         }
-        if (event.getNewState().getBlock() == Blocks.COBBLESTONE)
+
+        if (!ConfigTFC.GENERAL.disableCobbleGenOverride && event.getNewState().getBlock() == Blocks.COBBLESTONE)
```
This is already fixed in the current 1.12.x branch, and I believe this is showing up as a conflict. For reference it's fixed [here](https://github.com/TerraFirmaCraft/TerraFirmaCraft/blob/60a058d27865be5f3535c89438e5c209ec44b3d7/src/main/java/net/dries007/tfc/CommonEventHandler.java#L681)

alcatrazEscapee commented on src/main/java/net/dries007/tfc/ConfigTFC.java:
```diff
@@ -55,6 +55,21 @@ public static void onConfigChangedEvent(ConfigChangedEvent.OnConfigChangedEvent
+        @Config.Comment({"Disable the override TFC does when lava and water meet.",
+            "Only use if you want vanilla cobblestone or have another mod that changes this."})
+        @Config.LangKey("config." + MOD_ID + ".general.disableCobbleGenOverride")
+        public boolean disableCobbleGenOverride = false;
```
Same as previous, this is already fixed and this would be duplicating a config option.

alcatrazEscapee commented on src/main/java/net/dries007/tfc/api/capability/heat/CapabilityItemHeat.java:
```diff
@@ -39,6 +40,7 @@ public static void preInit()
         CapabilityManager.INSTANCE.register(IItemHeat.class, new DumbStorage<>(), ItemHeatHandler::new);
         //register heat on vanilla egg for cooking
         CapabilityItemHeat.CUSTOM_ITEMS.put(IIngredient.of(Items.EGG), () -> new ItemHeatHandler(null, 1, 480));
+        CapabilityItemHeat.CUSTOM_ITEMS.put(IIngredient.of(ItemsTFC.UNFIRED_BRICK), () -> new ItemHeatHandler(null, 1, 1601));
```
I don't believe this is used anymore since the unfired brick uses `ItemPottery`?

alcatrazEscapee commented on src/main/java/net/dries007/tfc/objects/blocks/BlocksTFC.java:
```diff
@@ -56,6 +60,114 @@
+    @GameRegistry.ObjectHolder("aggregate")
+    public static final BlockGravel AGGREGATE = getNull();
+
+    @GameRegistry.ObjectHolder("alabaster/bricks/white")
```
This is not the example of what's been done in other places in TFC, but I think it's both the best solution for this particular case, and what's most like the `RegistryObject` map based stuff I think we should move to in 1.15. So instead:
```java
public static final Map<EnumDyeColor, BlockDecorativeStone> ALABASTER_BRICKS = new HashMap<>();
public static final Map<EnumDyeColor, BlockDecorativeStone> ALABASTER_POLISHED = new HashMap<>();
public static final Map<EnumDyeColor, BlockDecorativeStone> ALABASTER_RAW = new HashMap<>();
```

alcatrazEscapee commented on src/main/java/net/dries007/tfc/objects/blocks/BlocksTFC.java:
```diff
@@ -115,6 +227,10 @@
+    private static ImmutableList<ItemBlock> allAlabasterRawBlocks;
+    private static ImmutableList<ItemBlock> allAlabasterBricksBlocks;
+    private static ImmutableList<ItemBlock> allAlabasterPolishedBlocks;
+
```
These shouldn't be necessary, the item blocks are all simple items and can have their model registration bundled with everything else.
```suggestion
```

alcatrazEscapee commented on src/main/java/net/dries007/tfc/objects/blocks/BlocksTFC.java:
```diff
@@ -266,6 +382,21 @@
+    {
+        return allAlabasterPolishedBlocks;
+    }
+
```
```suggestion
```

alcatrazEscapee commented on src/main/java/net/dries007/tfc/objects/blocks/BlocksTFC.java:
```diff
@@ -297,6 +431,37 @@ public static void registerBlocks(RegistryEvent.Register<Block> event)
 
         normalItemBlocks.add(new ItemBlockLargeVessel(register(r, "ceramics/fired/large_vessel", new BlockLargeVessel(), CT_POTTERY)));
+        {
+            Builder<ItemBlock> raw = ImmutableList.builder();
```
There's no need for the separate builders here. All the registration calls can be be adjusted to add the block to the respective map:
```java
// Inside each EnumDyeColor loop
BlockDecorativeStone block = new BlockDecorativeStone( ... );
normalItemBlocks.add(new ItemBlockTFC(register(r, "alabaster/whatever" + dyeColor.getName(), block)));
ALABASTER_RAW.put(color, block);
```

alcatrazEscapee commented on src/main/java/net/dries007/tfc/objects/blocks/BlocksTFC.java:
```diff
@@ -297,6 +431,37 @@ public static void registerBlocks(RegistryEvent.Register<Block> event)
+            raw.add(new ItemBlockTFC(register(r, "alabaster/raw/plain", new BlockDecorativeStone(MapColor.SNOW), CT_DECORATIONS)));
+            for (EnumDyeColor dyeColor : EnumDyeColor.values())
+            {
+                raw.add(new ItemBlockTFC(register(r, "alabaster/raw/" + dyeColor.getDyeColorName(), new BlockDecorativeStone(MapColor.getBlockColor(dyeColor)), CT_DECORATIONS)));
```
`EnumDyeColor#getDyeColorName` is annotated as `SideOnly(Side.CLIENT)`. Use `getName` instead, as it returns the same field.

alcatrazEscapee commented on src/main/java/net/dries007/tfc/objects/fluids/FluidsTFC.java:
```diff
@@ -64,6 +64,23 @@
     public static FluidWrapper BEER;
     public static FluidWrapper RUM;
+    // Dye
+    public static FluidWrapper BLACK_DYE;
```
Instead of this, we should mimic metal fluids, with a `Map<EnumDyeColor, FluidWrapper>`. I'm not concerned with making everything immutable as imo it makes more ugly code than it helps.

```java
// Storage
private static Map<EnumDyeColor, FluidWrapper> DYE;

public static FluidWrapper getFluidFromDye(EnumDyeColor color) { ... }
```

alcatrazEscapee commented on src/main/java/net/dries007/tfc/objects/fluids/FluidsTFC.java:
```diff
@@ -162,7 +179,24 @@ public static void registerFluids()
+                BROWN_DYE = registerFluid(new Fluid("brown_dye", STILL, FLOW, 0xFF835432)),
+                GREEN_DYE = registerFluid(new Fluid("green_dye", STILL, FLOW, 0xFF5E7C16)),
+                RED_DYE = registerFluid(new Fluid("red_dye", STILL, FLOW, 0xFFB02E26)),
+                BLACK_DYE = registerFluid(new Fluid("black_dye", STILL, FLOW, 0xFF1D1D21))
```
This can be replaced with a single map constructor.
 - One niche thing is that `EnumDyeColor#getColorValue` is client only, but `getColorComponentValues` is not. I think we can solve this via `new Color(f, g, b).getRGB()`

```java
DYE_FLUIDS = Arrays.stream(EnumDyeColor.values()).collect(Collectors.toMap(
    color -> color,
    color -> {
        float[] c = color.getColorComponentValues();
        return registerFluid(new Fluid(color.getName() + "_dye", STILL, FLOW, new Color(c[0], c[1], c[2]).getRGB()));
));
```

alcatrazEscapee commented on src/main/java/net/dries007/tfc/objects/items/metal/ItemIngot.java:
```diff
@@ -84,7 +85,7 @@ public EnumActionResult onItemUse(EntityPlayer player, World worldIn, BlockPos p
                         }
                     }
-                    else if (stateTop.getBlock().isReplaceable(worldIn, posTop) && worldIn.mayPlace(BlocksTFC.INGOT_PILE, posTop, false, EnumFacing.UP, null))
+                    else if (stateTop.getBlock().isReplaceable(worldIn, posTop) && worldIn.mayPlace(BlocksTFC.INGOT_PILE, posTop, false, EnumFacing.UP, null) && worldIn.getBlockState(posTop).getBlock().getBlockFaceShape(worldIn, worldIn.getBlockState(posTop), posTop, EnumFacing.UP) == BlockFaceShape.SOLID)
```
```suggestion
                    else if (stateTop.getBlock().isReplaceable(worldIn, posTop) && worldIn.mayPlace(BlocksTFC.INGOT_PILE, posTop, false, EnumFacing.UP, null) && stateTop.getBlockFaceShape(worldIn, posTop, EnumFacing.UP) == BlockFaceShape.SOLID)
```

alcatrazEscapee commented on src/main/java/net/dries007/tfc/types/DefaultRecipes.java:
```diff
@@ -134,10 +138,168 @@ public static void onRegisterBarrelRecipeEvent(RegistryEvent.Register<BarrelReci
         );
+
+        //Nice handy list of dye in the correct order BY META DATA to get the correct fluid.
+        List<FluidWrapper> fluids = new ArrayList<>();
```
No please. If the fluids are stored via `Map<EnumDyeColor, FluidWrapper>`, the meta data can be retrieved automatically via `EnumDyeColor#getMetadata()`

i.e. `fluids.get(dyeMeta).get()` gets replaced wth `FluidsTFC.DYE.get(color).get()`

alcatrazEscapee commented on src/main/java/net/dries007/tfc/types/DefaultRecipes.java:
```diff
@@ -134,10 +138,168 @@ public static void onRegisterBarrelRecipeEvent(RegistryEvent.Register<BarrelReci
+        //The many many many recipes that is dye. This assumes that the standard meta values for colored objects are followed.
+        for (EnumDyeColor dyeColor : EnumDyeColor.values())
+        {
+            String dyeName = dyeColor.getDyeColorName();
```
`EnumDyeColor#getDyeColorName` is client only, don't use.
```suggestion
            String dyeName = dyeColor.getName();
            if (dyeColor == EnumDyeColor.SILVER)
            {
                dyeName = "light_gray";
            }
```

alcatrazEscapee commented on src/main/java/net/dries007/tfc/types/DefaultRecipes.java:
```diff
@@ -134,10 +138,168 @@ public static void onRegisterBarrelRecipeEvent(RegistryEvent.Register<BarrelReci
+                    new BarrelRecipe(IIngredient.of(fluids.get(dyeMeta).get(), 125), IIngredient.of("powderConcrete"), null, new ItemStack(Blocks.CONCRETE_POWDER, 1, dyeMeta), ICalendar.TICKS_IN_HOUR).setRegistryName("concrete_" + dyeName),
+                    new BarrelRecipe(IIngredient.of(fluids.get(dyeMeta).get(), 125), IIngredient.of(BlocksTFC.AGGREGATE), null, new ItemStack(Blocks.CONCRETE_POWDER, 1, dyeMeta), ICalendar.TICKS_IN_HOUR).setRegistryName("aggregate_" + dyeName)
+                );
+            }
```
Don't need to duplicate all these recipe names, just change the dye name initially and call the registration the way.

alcatrazEscapee commented on src/main/java/net/dries007/tfc/types/DefaultRecipes.java:
```diff
@@ -134,10 +138,168 @@ public static void onRegisterBarrelRecipeEvent(RegistryEvent.Register<BarrelReci
+            }
+        }
+
+        event.getRegistry().registerAll(
```
This should be three lines in a for loop
```java
for (EnumDyeColor color : EnumDyeColor.values())
{
    event.getRegistry().registerAll(
        new BarrelRecipe(IIngredient.of(FluidsTFC.DYE.get(color).get(), 125), IIngredient.of("alabasterBricks"), null, new ItemStack(BlocksTFC.ALABASTER_BRICKS.get(color)), ICalendar.TICKS_IN_HOUR).setRegistryName("alabaster_bricks_" + color.getName()),
        // etc.
    );
}
```

alcatrazEscapee reviewed (changes requested):
Initial feedback on where code cleanup could happen. Whack me if questions.

Claycorp committed:
Fixed grass landing particulates not being made of grass. (Closes #710)
Fixed all the issues alc pointed out I think and they should still work.
Also fixed the ingot placing to work correctly instead of only stacking two blocks tall.

Claycorp commented:
Also this fixed #891 Forgot to add that to the commits.
